Mathlib/FieldTheory/JacobsonNoether.lean

```diff
@@ -0,0 +1,327 @@
+      · exact Nat.sub_one_lt <| ne_of_gt (Nat.find_spec exist).1
+    · rw [not_lt, Nat.le_zero] at choice
+      have := Nat.eq_add_of_sub_eq (Nat.find_spec exist).1 ht.symm
+      simp only [gt_iff_lt, choice, Nat.succ_eq_add_one, zero_add] at this
+      rw [this, pow_one]
+      exact hb1
+    -- We define `c` to be the value that we proved above to be non-zero.
+  set c := ((δ a) ^ n) b with hc_def
+  letI : Invertible c := ⟨c⁻¹, inv_mul_cancel₀ (hb.1), mul_inv_cancel₀ (hb.1)⟩
+  -- We prove that `c` is commute with `a`.
+  have hc : c * a = a * c := by
+    symm; apply eq_of_sub_eq_zero
+    rw [← LinearMap.mulLeft_apply (R := k), ← LinearMap.mulRight_apply (R := k),
+      ← δ_def, δ_iterate_succ a b n, hb.2]
+  -- We now make some computation to obtain the final equation.
+  set d := c⁻¹ * a * ((δ a) ^ (n - 1)) b with hd_def
+  have hc': c⁻¹ * a = a * c⁻¹ := by
+    apply_fun (c⁻¹ * · * c⁻¹) at hc
+    rw [← mul_assoc, inv_mul_cancel₀ hb.1, one_mul, mul_assoc, mul_assoc,
+      mul_inv_cancel₀ hb.1, mul_one] at hc
+    exact hc.symm
+  have c_eq : a * ((δ a) ^ (n - 1)) b - ((δ a) ^ (n - 1)) b * a = c := by
+    rw [hc_def, ← Nat.sub_add_cancel hn, ← δ_iterate_succ, δ_def]; rfl
+  have eq1 : c⁻¹ * a * ((δ a)^ (n - 1)) b - c⁻¹ * ((δ a) ^ (n - 1)) b * a = 1 := by
+    simp_rw [mul_assoc, (mul_sub_left_distrib c⁻¹ _ _).symm, c_eq, inv_mul_cancel_of_invertible]
+  -- We show that `a` is commute with `d`.
+  have deq : a * d - d * a = a := by
+    nth_rw 3 [← mul_one a]
+    rw [hd_def, ← eq1, mul_sub, mul_assoc _ _ a, sub_right_inj, hc',
+      ← mul_assoc, ← mul_assoc, ← mul_assoc]
+  -- This then derives a contradiction.
+  apply_fun (a⁻¹ * · ) at deq
+  rw [mul_sub, ← mul_assoc, inv_mul_cancel₀ ha₀, one_mul, ← mul_assoc, sub_eq_iff_eq_add] at deq
+  obtain ⟨r, hr⟩ := (exists_pow_mem_center_ofInseparable p d hinsep)
+  apply_fun (· ^ (p ^ r)) at deq
+  rw [add_pow_expChar_pow_of_commute D 1 _ (Commute.one_left _) , one_pow,
+    ← conj_nonComm_Algebra (ha := ha₀), ← hr.comm, mul_assoc, inv_mul_cancel₀ ha₀, mul_one,
+    self_eq_add_left] at deq
+  exact one_ne_zero deq
+
+end JacobsonNoether
+
+namespace canary
+
+open Algebra Subring Polynomial
+
+example {L D : Type*} [Field L] [DivisionRing D] [Algebra L D] [alg : Algebra.IsAlgebraic L D]
+    (hcenter : Subalgebra.center L D = ⊥) (hneq : (⊥ : Subalgebra L D) ≠ ⊤) :
+    ∃ x : D, x ∉ (⊥ : Subalgebra L D) ∧ IsSeparable L x := by
+  set k' := (⊥ : Subalgebra L D).toSubring with hk'
+  have hcenter' : k' = center D := by apply (hcenter.symm) ▸ hk'
+  have ntrivial : k' ≠ ⊤ := Ne.intro fun heq ↦ hneq <| toSubring_eq_top.mp <| hk' ▸ heq
+  set ψ : L ≃+* (⊥ : Subalgebra L D) := ⟨((Algebra.botEquiv L D).toRingEquiv).symm,
+      fun _ _ ↦ by simp only [Equiv.toFun_as_coe, EquivLike.coe_coe, map_mul],
+      fun _ _ ↦ by simp only [Equiv.toFun_as_coe, EquivLike.coe_coe, map_add]⟩
+  set φ : (⊥ : Subalgebra L D) ≃+* (center D) := hcenter ▸ RingEquiv.refl (Subalgebra.center L D)
```

Comment:
It's best to use `Subalgebra.equivOfEq` here. The `▸` probably will erase the actual definition of `φ`.

Reply:
Indeed that make it easier to reach the definition, thanks!